pumpkin-world/src/world_gen/noise/perlin.rs

```diff
@@ -0,0 +1,1134 @@
+    }
+
+    fn grad(hash: i32, x: f64, y: f64, z: f64) -> f64 {
+        dot(&GRADIENTS[(hash & 15) as usize], x, y, z)
+    }
+
+    fn perlin_fade(value: f64) -> f64 {
+        value * value * value * (value * (value * 6f64 - 15f64) + 10f64)
+    }
+
+    fn map(&self, input: i32) -> i32 {
+        self.permutation[(input & 0xFF) as usize] as i32
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn sample(
+        &self,
+        x: i32,
+        y: i32,
+        z: i32,
+        local_x: f64,
+        local_y: f64,
+        local_z: f64,
+        fade_local_y: f64,
+    ) -> f64 {
+        let i = self.map(x);
+        let j = self.map(x.wrapping_add(1));
+        let k = self.map(i.wrapping_add(y));
+
+        let l = self.map(i.wrapping_add(y).wrapping_add(1));
+        let m = self.map(j.wrapping_add(y));
+        let n = self.map(j.wrapping_add(y).wrapping_add(1));
+
+        let d = Self::grad(self.map(k.wrapping_add(z)), local_x, local_y, local_z);
+        let e = Self::grad(
+            self.map(m.wrapping_add(z)),
+            local_x - 1f64,
+            local_y,
+            local_z,
+        );
+        let f = Self::grad(
+            self.map(l.wrapping_add(z)),
+            local_x,
+            local_y - 1f64,
+            local_z,
+        );
+        let g = Self::grad(
+            self.map(n.wrapping_add(z)),
+            local_x - 1f64,
+            local_y - 1f64,
+            local_z,
+        );
+        let h = Self::grad(
+            self.map(k.wrapping_add(z).wrapping_add(1)),
+            local_x,
+            local_y,
+            local_z - 1f64,
+        );
+        let o = Self::grad(
+            self.map(m.wrapping_add(z).wrapping_add(1)),
+            local_x - 1f64,
+            local_y,
+            local_z - 1f64,
+        );
+        let p = Self::grad(
+            self.map(l.wrapping_add(z).wrapping_add(1)),
+            local_x,
+            local_y - 1f64,
+            local_z - 1f64,
+        );
+        let q = Self::grad(
+            self.map(n.wrapping_add(z).wrapping_add(1)),
+            local_x - 1f64,
+            local_y - 1f64,
+            local_z - 1f64,
+        );
+        let r = Self::perlin_fade(local_x);
+        let s = Self::perlin_fade(fade_local_y);
+        let t = Self::perlin_fade(local_z);
+
+        lerp3(r, s, t, d, e, f, g, h, o, p, q)
+    }
+}
+
+pub struct OctavePerlinNoiseSampler {
+    octave_samplers: Vec<Option<PerlinNoiseSampler>>,
+    amplitudes: Vec<f64>,
+    #[allow(dead_code)]
+    first_octave: i32,
+    persistence: f64,
+    lacunarity: f64,
+    max_value: f64,
+}
+
+impl OctavePerlinNoiseSampler {
+    fn get_total_amplitude(scale: f64, persistence: f64, amplitudes: &[f64]) -> f64 {
+        let mut d = 0f64;
+        let mut e = persistence;
+
+        for amplitude in amplitudes.iter() {
+            if *amplitude != 0f64 {
+                d += amplitude * scale * e;
+            }
```

Comment:
The `if` statement is useless and 

```suggestion
            d += amplitude * scale * e;
```
```
branch               time:   [24.947 ns 25.014 ns 25.127 ns]

brancheless               time:   [23.428 ns 23.442 ns 23.454 ns]
                        change: [-6.8377% -6.3385% -5.9932%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Reply:
will update